### ISSUE TITLE
Discussion about less strict code of the TinylogLoggingProvider

### ISSUE THREAD

**Issue by Git5000** [tinylog-org/tinylog#168]
**Please see this whole feature request just as basis for disussion**. It requires more in depth code changes as before (but no API or user experience change) so it is only a suggestion and basis for discussion:

**Goal:** Allow easier user defined LoggingProviders in particular resusing code of the `TinylogLoggingProvider`. Since `TinylogLoggingProvider` is very private and static one has to copy all code for `TinylogLoggingProvider `and also `ConfigurationParser`. The goal would be to move some general code out of the fixed classes to allow subclassing and resuing them.

Since in some other feature request on this project, a flexible configuration was rejected for performance reasons it is important that the performance of the logger framework does not suffer. However, all code changes apply only to the construction of the logging provider not its use. So some more flexibility there should not impact overall performance

**Note:** All here and in the following "compare" is only a quick hack and only meant as basis for discussion. Here you see a compare V2.2 code to new suggested code in my test repo https://github.com/Git5000/tinylog/compare/v2.2...git5001:v2.2

**Change overview:**

**TinylogLoggingProvider**:
  - use 1 or 2 configuration object(s) in constructor instead of the static `ConfigurationParser`
  - move static methods to this or these configuration objects e.g. called `TinylogLoggingConfiguration `as static and/or 
    non static    (e.g. `calculateRequiredLogEntryValues `or `static createLogEntry`)
  - A bit unrelated but: Is there a good reason for removing "tag" and "level" attributes from the writer creation property array
    can be useful to have them? 

**TinylogLoggingConfiguration**: A new non final class with mainly non static methods to allow subclassing.
  - collects the methods from `TinylogLoggingProvider `and `ConfigurationParser` as non static. 
  - Optionally, this can be all inside `ConfigurationParser` - then only one class is needed. But the name configuration parser
    indicates that it does just handle the parsing not a lot more e.g. not create the writers etc But if onyl one class is desirend one can combine it in `ConfigurationParser `and/or rename it.

**ConfigurationParser**: Make class non final and methods non static
  - create local / temporariy object of this in the `TinylogLoggingProvider `constructor
  - move writer related code to new class TinylogLoggingConfiguration
 - With above two changes the `ConfigurationParser `can also stay static as it has less functionality. However, it would fit to make it non static.

**Very optional:** 

- make `TinylogLoggingProvider `protected instead of private to allow sublcassing (I have a feeling that this might not be desired). However, above changes alone allow a very small new logging provider using the new configuration classes. So this is not really nessesary anymore.

**Comment by pmwmedia** [comment]
Thank you very much for your detailed feature request! It sounds very promising :) I agree that it should become easier to implement a custom logging provider. In general, we should get rid of final and static, where it isn't necessary for performance or usability reasons. I really like your idea of introducing `TinylogLoggingConfiguration` and making `ConfigurationParser` non static.

> A bit unrelated but: Is there a good reason for removing "tag" and "level" attributes from the writer creation property array
can be useful to have them?

I can't find any good reason for removing these attributes.  So, I think these code for removing these attributes can be removed.

> Make TinylogLoggingProvider protected instead of private to allow subclassing (I have a feeling that this might not be desired).

In general, it would be okay for as long as the [benchmark](https://tinylog.org/v2/benchmark/) won't suffer from such a change. We need to be sure that the JVM will be still able to inline small methods and remove dead code based on the configured logging level. These tricks make an huge performance difference.

**Comment by Git5000** [comment]
Super. I will look at the benchmarks and see what has an effect and what not. I keep you updated (but it might take a few days).

**Comment by pmwmedia** [comment]
I'm looking forward to your pull request :)

This a big change of course. No rush! As tinylog 2.2 is already close to release, I see your change in version 2.3.

**Comment by git5001** [comment]
I get the following output from the (original unchanged) tinylog build benchmark. Could you tell me what I need to pay attention to when testing for changes. So what should not go lower?

```
Result "org.tinylog.benchmarks.logging.Tinylog2Benchmark.output":
  259732,486 ±(99.9%) 16548,822 ops/s [Average]
  (min, avg, max) = (95082,433, 259732,486, 297417,209), stdev = 48794,579
  CI (99.9%): [243183,663, 276281,308] (assumes normal distribution)

Benchmark                  (async)   Mode  Cnt           Score           Error  Units
Tinylog2Benchmark.discard    false  thrpt  100  1083361217,760 ± 100606775,882  ops/s
Tinylog2Benchmark.discard     true  thrpt  100  1114063915,785 ± 114632426,080  ops/s
Tinylog2Benchmark.output     false  thrpt  100       82985,660 ±      8243,572  ops/s
Tinylog2Benchmark.output      true  thrpt  100      259732,486 ±     16548,822  ops/s
```

**Comment by Git5000** [comment]
I get the following output from the (original unchanged) tinylog build benchmark. Could you tell me what I need to pay attention to when testing for changes. So what should not go lower?

```
Result "org.tinylog.benchmarks.logging.Tinylog2Benchmark.output":
  259732,486 ±(99.9%) 16548,822 ops/s [Average]
  (min, avg, max) = (95082,433, 259732,486, 297417,209), stdev = 48794,579
  CI (99.9%): [243183,663, 276281,308] (assumes normal distribution)

Benchmark                  (async)   Mode  Cnt           Score           Error  Units
Tinylog2Benchmark.discard    false  thrpt  100  1083361217,760 ± 100606775,882  ops/s
Tinylog2Benchmark.discard     true  thrpt  100  1114063915,785 ± 114632426,080  ops/s
Tinylog2Benchmark.output     false  thrpt  100       82985,660 ±      8243,572  ops/s
Tinylog2Benchmark.output      true  thrpt  100      259732,486 ±     16548,822  ops/s
```

PS: Posted again with the right GITHUB account...

**Comment by pmwmedia** [comment]
Actually, the score (log entries per second) of all of these four benchmarks shouldn't decrease.

**Comment by Git5000** [comment]
I finally managed to run the benchmarks "original branch V2.2" versus the "proposed change #168". See below for the results. I think they can be considered the same considering measurement errors.  You probably know better how accurate the results are. Since the new version is measuring a slightly higher score (which actually cannot be) I think this is all inside the measurement errors.


**Original V2.2:**
```
Benchmark                  (async)   Mode  Cnt           Score         Error  Units
Tinylog2Benchmark.discard    false  thrpt  100  2297887080,036 ± 2526052,794  ops/s
Tinylog2Benchmark.discard     true  thrpt  100  2303772632,094 ± 2851738,447  ops/s
Tinylog2Benchmark.output     false  thrpt  100      190461,047 ±    2408,265  ops/s
Tinylog2Benchmark.output      true  thrpt  100      363798,345 ±    5810,157  ops/s
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  02:13 h
```


**Modified Git5000 Issue #168:**
```
Benchmark                  (async)   Mode  Cnt           Score         Error  Units
Tinylog2Benchmark.discard    false  thrpt  100  2336129231,891 ± 5153167,424  ops/s
Tinylog2Benchmark.discard     true  thrpt  100  2359598293,701 ± 6390528,418  ops/s
Tinylog2Benchmark.output     false  thrpt  100      208570,339 ±    2055,706  ops/s
Tinylog2Benchmark.output      true  thrpt  100      407206,758 ±    1068,143  ops/s
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  02:13 h
```

**Comment by pmwmedia** [comment]
I agree, there seems to be no negative impact on the performance due to your changes :-) Looks very promising!

Maybe you found a performance tweak by accident or there were any side effects like hotter room temperatures during the first run. What ever was the reason, you have proofed that your code has no negative impact on the performance.

**Comment by Git5000** [comment]
I think this issue can be closed now.

**Comment by pmwmedia** [comment]
Fixed by #173
